src/erlcloud_ec2.erl

```diff
@@ -3574,6 +3577,142 @@ ec2_query(Config, Action, Params, ApiVersion) ->
                                   Config#aws_config.ec2_port,
                                   "/", QParams, "ec2", Config).
 
+
+% Exported Query Function with parameter handling
+
+-spec query(aws_config(), string(), map()) -> ok_error().
+query(Config, Action, Params) when is_map(Params) ->
+    query(Config, Action, Params, [], none, ?API_VERSION).
+
+-spec query(aws_config(), string(), map(), list() | atom()) -> ok_error().
+query(Config, Action, Params, Filter) when is_map(Params), is_list(Filter) ->
+    query(Config, Action, Params, Filter, none, ?API_VERSION);
+query(Config, Action, Params, ResponseFormat) when is_map(Params), is_atom(ResponseFormat) ->
+    query(Config, Action, Params, [], ResponseFormat, ?API_VERSION).
+
+-spec query(aws_config(), string(), map(), list(), atom()|string()) -> ok_error().
+query(Config, Action, Params, Filter, ResponseFormat) when is_map(Params), is_atom(ResponseFormat), is_list(Filter) ->
+    query(Config, Action, Params, Filter, ResponseFormat, ?API_VERSION);
+query(Config, Action, Params, Filter, ApiVersion) when is_map(Params), is_list(Filter)->
+    query(Config, Action, Params, Filter, none, ApiVersion).
+
+-spec query(aws_config(), string(), map(), list(), atom(), string()) -> ok_error().
+query(Config, Action, Params, Filter, ResponseFormat, ApiVersion) when is_map(Params), is_atom(ResponseFormat) ->
+    parse_response(do_query(Config, Action, Params, Filter, ApiVersion), ResponseFormat).
+
+do_query(Config, Action, MapParams, Filter, ApiVersion) -> 
+    Params = prepare_action_params(MapParams, Filter),
+    case ec2_query(Config, Action, Params, ApiVersion) of
+        {ok, Results} ->
+            {ok, Results};
+        {error, _} = E -> E
+    end.
+
+parse_response({ok, Response}, map) ->
+    {ok, _Res} = erlcloud_xml:xml_to_map(Response);
+parse_response({ok, Response}, _) ->
+    {ok, Response};
+parse_response({error, _} = ErrRes, _Format) -> 
+    ErrRes.
+
+%%%%
+prepare_action_params(ParamsMap, []) when is_map(ParamsMap) ->
+    flatten(map_to_kv(ParamsMap));
+prepare_action_params(ParamsMap, Filters) when is_map(ParamsMap) ->
+    % Convert to list of tuples, with {Key, Value} format for each record
+    % Flatten the list of tuples
+    FlatList = flatten(map_to_kv(ParamsMap)),
+    FlatList ++ list_to_ec2_filter(Filters). % Add the filters 
+
+map_to_kv(Map) when is_map(Map) ->
```

Comment:
I think you can transform and flatten in one go if extend map_to_kv with one more argument.